File: js/src/codec.ts
import { copyBytes, textDecoder, textEncoder, toUint8Array, type Bytes } from "./bytes.js";

export interface Codec<T = unknown> {
  encode(value: T): Uint8Array;
  decode(bytes: Bytes): T;
}

export const jsonCodec: Codec = {
  encode(value: unknown): Uint8Array {
    const json = JSON.stringify(value);
    return textEncoder.encode(json === undefined ? "null" : json);
  },

  decode<T = unknown>(bytes: Bytes): T {
    const data = toUint8Array(bytes);
    if (data.byteLength === 0) {
      return undefined as T;
    }
    return JSON.parse(textDecoder.decode(data)) as T;
  },
};

export const stringCodec: Codec<string> = {
  encode(value: string): Uint8Array {
    return textEncoder.encode(value);
  },

  decode(bytes: Bytes): string {
    return textDecoder.decode(toUint8Array(bytes));
  },
};

export const bytesCodec: Codec<Bytes> = {
  encode(value: Bytes): Uint8Array {
    return copyBytes(value);
  },

  decode(bytes: Bytes): Uint8Array {
    return copyBytes(bytes);
  },
};

File: python/rpc_core/codec.py
from __future__ import annotations

import json
from typing import Any, Generic, Protocol, TypeVar

T = TypeVar("T")


class Codec(Protocol[T]):
    def encode(self, value: T) -> bytes:
        ...

    def decode(self, data: bytes | bytearray | memoryview) -> T:
        ...


class JsonCodec(Generic[T]):
    def encode(self, value: T) -> bytes:
        return json.dumps(value, ensure_ascii=False, separators=(",", ":")).encode("utf-8")

    def decode(self, data: bytes | bytearray | memoryview) -> Any:
        payload = bytes(data)
        if not payload:
            return None
        return json.loads(payload.decode("utf-8"))


class StringCodec:
    def encode(self, value: str) -> bytes:
        return value.encode("utf-8")

    def decode(self, data: bytes | bytearray | memoryview) -> str:
        return bytes(data).decode("utf-8")


class BytesCodec:
    def encode(self, value: bytes | bytearray | memoryview) -> bytes:
        return bytes(value)

    def decode(self, data: bytes | bytearray | memoryview) -> bytes:
        return bytes(data)


json_codec: Codec[Any] = JsonCodec()
string_codec = StringCodec()
bytes_codec = BytesCodec()
